pkg/server/bootstrap/api-export-identity_controller.go

```diff
@@ -0,0 +1,255 @@
+				return false
+			}
+			clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+			if clusterName != configshard.SystemShardCluster {
+				return false
+			}
+			switch t := obj.(type) {
+			case *corev1.ConfigMap:
+				return t.Name == configMapName
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.queue.Add(workKey) },
+			UpdateFunc: func(old, new interface{}) { c.queue.Add(workKey) },
+			DeleteFunc: func(obj interface{}) { c.queue.Add(workKey) },
+		},
+	})
+	return c, nil
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+	defer klog.Infof("Shutting down %s controller", ControllerName)
+
+	go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.reconcile(ctx)
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *controller) reconcile(ctx context.Context) error {
+	rawApiExports, err := c.remoteShardApiExportsIndexer.ByIndex(indexers.ByLogicalCluster, tenancyv1alpha1.RootCluster.String())
+	if err != nil {
+		return err
+	}
+	interestingApiExports := map[string]*apisv1alpha1.APIExport{}
+	for _, rawApiExport := range rawApiExports {
+		apiExport := rawApiExport.(*apisv1alpha1.APIExport)
+		if !c.interestingAPIExportNames.Has(apiExport.Name) {
+			continue
+		}
+		if apiExport.Status.IdentityHash == "" {
+			return fmt.Errorf("apiExport %s is not ready", apiExport.Name)
+		}
+		interestingApiExports[apiExport.Name] = apiExport
+	}
+
+	requiredApiExportIdentitiesConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      configMapName,
+		},
+		Data: map[string]string{},
+	}
+	for _, name := range c.groupExportNames {
+		apiExport, found := interestingApiExports[name]
+		if !found {
+			return fmt.Errorf("required %s apiExport hasn't been found", name)
+		}
+		requiredApiExportIdentitiesConfigMap.Data[name] = apiExport.Status.IdentityHash
+	}
+	for _, name := range c.groupResourceExportNames {
+		apiExport, found := interestingApiExports[name]
+		if !found {
+			return fmt.Errorf("required %s apiExport hasn't been found", name)
+		}
+		requiredApiExportIdentitiesConfigMap.Data[name] = apiExport.Status.IdentityHash
+	}
+
+	rawConfigMaps, err := c.configMapIndexer.ByIndex(indexers.ByLogicalClusterAndNamespace, clusters.ToClusterAwareKey(configshard.SystemShardCluster, "default"))
```

Comment:
What's the meaning of listing everything in the namespace, then iterating to find the one name we need, as opposed to just getting it from the lister?

Reply:
ah, right, I've just realized I could've used `clusters.ToClusterAwareKey` to construct the name.